src/common/ad_tx.c

```diff
@@ -608,3 +613,423 @@ ad_tx_redo_act_next(struct ad_tx *tx)
 	}
 	return &tx->tx_redo_act_pos->it_act;
 }
+
+static struct ad_tx *
+tx_get()
+{
+	return tls_tx;
+}
+
+static void
+tx_set(struct ad_tx *tx)
+{
+	tls_tx = tx;
+}
+
+static void
+tx_callback(struct ad_tx *tx)
+{
+	if (!tx->tx_stage_cb || tx->tx_layer != 0)
+		return;
+	tx->tx_stage_cb(ad_tx_stage(tx), tx->tx_stage_cb_arg);
+}
+
+static int
+tx_end(struct ad_tx *tx, int err)
```

Comment:
Should we make it a generic function so that it could be shared by BMEM as well?